elastic/eui
# [EuiSuggest][EuiSelectable] Allow controlled search `value` prop

### Summary

closes https://github.com/elastic/eui/issues/5644
closes https://github.com/elastic/eui/issues/5651
closes https://github.com/elastic/eui/issues/4341 - value, disabled, and placeholder can now all be passed without Typescript issues

- **Enhancements**
  - Added the ability for `EuiSelectable` and `EuiSuggest` to accept controlled `value` props
- **Bug fixes**
  - Fixed `EuiSuggest` not correctly passing props to the search input (53c328e)
  - Fixed `EuiSelectable` incorrectly rendering the passed `id` prop on the listbox instead of the parent wrapper (809ec87)
    - Rendering the parent ID on the listbox is problematic because the listbox is conditionally rendered when searchable. If no results are returned, the ID disappears from the page, which isn't great developer experience.
  - Fixed `EuiSelectable` to no longer call `searchProps.onChange` when list items are clicked (b969621)
  - Fixed `EuiSelectable` not respecting `searchProps.inputRef` (514b8f9)
- **Breaking changes**
    - Removed `EuiSelectable`'s `searchProps.onSearch` prop (since Enter keypresses do not trigger a search callback) - use `searchProps.onChange` instead
    - Renamed `EuiSuggest`'s `onInputChange` and `onSearchChange` callbacks to `onInput`/`onSearch` respectively, for consistency with our existing callback naming conventions
    - Removed `EuiSuggest`'s `isLoading` prop - use `status.loading` instead
    - Removed the `incremental` prop from `EuiSuggest` and `EuiSelectable`'s `searchProps`
- **Refactors**
  - `EuiSelectableSearch` was converted from a class component to a functional component, and no longer maintains its own internal `searchValue` state (but inherits it from the parent `EuiSelectable`, which already maintains said state)

If possible, I recommend following along [by commit](https://github.com/elastic/eui/pull/5658/commits).

## QA

- EuiSuggest
  - Go to http://localhost:8030/#/forms/suggest
  - [x] Confirm the first example loads in with the `field` value, but still works as normal/expected when the user types or backspaces into the search input
  - [x] Go to the last example on the page (no controlled `value` prop) and type into the search input. Confirm that still works as normal/expected
- EuiSelectable
  - Go to http://localhost:8030/#/forms/selectable#searchable
  - [x] Confirm the `Searchable` example loads in with a `t` value, but still works as normal/expected when the user types or backspaces into the search input
  - [x] Go to the last example on the page (no controlled `value` prop) and type into the search input. Confirm that still works as normal/expected

### Checklist

~- [ ] Check against **all themes** for compatibility in both light and dark modes~
~- [ ] Checked in **mobile**~
~- [ ] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**~
~- [ ] Checked for **accessibility** including keyboard-only and screenreader modes~


~- [ ] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**~
~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
